index.js

```diff
@@ -8,297 +12,354 @@ const clearInterval = require("sdk/timers").clearInterval;
+            if (typeof price[jsonPath[i]] == "undefined") {
+              if (DEBUG) {
+                console.log("BitcoinPriceTicker error loading ticker " + id + 
+                            ". URL is not correctly responding:" + url);
+              }
+              return;
+            }
+            price = price[jsonPath[i]];
+          }
+          if (DEBUG) {
+            console.log("Price received and parsed: " + price);
+          }
+          tickersFrame.postMessage({
+            "type": "updateTickerModelPrice",
+            "id": id,
+            "data": {
+              "price": price
+            }
+          }, tickersFrame.url);
+        }
+      }
+    }).get();
+  }
+
+  function startAutoPriceUpdate(tickerId) {
+    if (tickers[tickerId].url && tickers[tickerId].jsonPath) {
+      var fetchURLDataWrapper = function() {
+        fetchURLData(tickerId, tickers[tickerId].url, tickers[tickerId].jsonPath);
+      };
+      fetchURLDataWrapper();
+      tickers[tickerId].timer = setInterval(fetchURLDataWrapper, (tickers[tickerId].updateInterval * 1000));
+    }
+  }
+
+  function stopAutoPriceUpdate(tickerId) {
+    if (tickers[tickerId].timer) { // Remove previous auto-update call if any
+      clearInterval(tickers[tickerId].timer); // Stop automatic refresh
+      tickers[tickerId].timer = null;
+    }
+  }
 
-  // Refresh ticker when changing add-on options
-  var updateTickerCaller = function(tickerName, onlyStyle) {
-    if ( (tickers[tickerName] != null) && (tickers[tickerName].length >= 2)) {
-      if (onlyStyle) {
-        tickers[tickerName][2](); // Call update Ticker method to update only the style
-      } else {
-        tickers[tickerName][1](); // Call update Ticker method
+  function updateTickerRefreshIntervalForTicker(tickerId) {
+    var refreshRate = getIntegerPreference("Timer");
+    if (refreshRate < 1) {
+      refreshRate = DEFAULT_REFRESH_RATE;
+    }
+    if (tickers[tickerId] && tickers[tickerId].enabled) {
+      if (DEBUG) {
+        console.log("updateTickerRefreshIntervalForTicker:" + tickerId);
+      }
+      if (tickers[tickerId].updateInterval != refreshRate) { // Update the real interval
+        tickers[tickerId].updateInterval = refreshRate;
+        stopAutoPriceUpdate(tickerId);
+        startAutoPriceUpdate(tickerId);
       }
     }
-  };
+  }
 
   // Create new refresh interval for each ticker when option is changed
-  var updateTickerRefreshInterval = function() {
-    var refresh_rate = getIntegerPreference("Timer");
-    if (refresh_rate < 1) {
-      refresh_rate = DEFAULT_REFRESH_RATE;
-    }
-    for (var i in tickers) {
-      if ( (tickers[i] != null) && (tickers[i].length >= 4)) {
-        clearInterval(tickers[i][3]); // Stop automatic refresh of removed ticker
-        tickers[i][3] = setInterval(tickers[i][1], (refresh_rate * 1000));
+  function updateTickerRefreshInterval() {
+    for (var tickerId in tickers) { // Update all tickers that require it
```

Comment:
The body of a for in should be wrapped in an if statement to filter unwanted properties from the prototype.